src/webui/api/torrentscontroller.cpp

```diff
@@ -679,6 +706,58 @@ void TorrentsController::removeTrackersAction()
         torrent->forceReannounce();
 }
 
+void TorrentsController::addPeersAction()
+{
+    checkParams({"hash", "peers"});
+
+    const QString hash = params()["hash"];
+    const QStringList peers = params()["peers"].split('|');
+
+    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QList<BitTorrent::PeerAddress> peerList;
+    for (const QString &peer : peers) {
+        const BitTorrent::PeerAddress addr = parsePeer(peer);
+        if (addr.ip.isNull())
+            throw APIError(APIErrorType::BadParams);
+        peerList.append(addr);
+    }
+
+    int peersAdded = 0;
+    for (const BitTorrent::PeerAddress &peer : peerList) {
+        if (torrent->connectPeer(peer))
+            ++peersAdded;
+    }
+    if (peersAdded == 0)
+        throw APIError(APIErrorType::Conflict, tr("No peers were added"));
```

Comment:
Is it necessary to throw an exception here? What's the point?

Reply:
Yes. But why this particular case is error?
This is only a special case of the fact that some peers can not be added. If (e.g.) 9 from 10 peers aren't added it's OK for you (not error).
Why not just tell the client the amount of added peers?